url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/dnssec/DNSSECClient.java

```diff
@@ -0,0 +1,555 @@
+        String[] childSplit = child.split("\\.");
+        String[] parentSplit = parent.split("\\.");
+        if (parentSplit.length > childSplit.length) return false;
+        for (int i = 1; i <= parentSplit.length; i++) {
+            if (!parentSplit[parentSplit.length - i].equals(childSplit[childSplit.length - i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private Set<UnverifiedReason> verifySignedRecords(Question q, RRSIG rrsig, List<Record<? extends Data>> records) throws IOException {
+        Set<UnverifiedReason> result = new HashSet<>();
+        DNSKEY dnskey = null;
+        if (rrsig.typeCovered == TYPE.DNSKEY) {
+            // Key must be present
+            for (Record<? extends Data> record : records) {
+                Record<DNSKEY> dnsKeyRecord = record.ifPossibleAs(DNSKEY.class);
+                if (dnsKeyRecord == null) continue;
+
+                if (dnsKeyRecord.payloadData.getKeyTag() == rrsig.keyTag) {
+                    dnskey = dnsKeyRecord.payloadData;
+                    break;
+                }
+            }
+        } else if (q.type == TYPE.DS && rrsig.signerName.equals(q.name)) {
+            // We should not probe for the self signed DS negative response, as it will be an endless loop.
+            result.add(new NoTrustAnchorReason(q.name.ace));
+            return result;
+        } else {
+            DNSSECMessage dnskeyRes = queryDnssec(rrsig.signerName, TYPE.DNSKEY);
+            if (dnskeyRes == null) {
+                throw new DNSSECValidationFailedException(q, "There is no DNSKEY " + rrsig.signerName + ", but it is used");
+            }
+            result.addAll(dnskeyRes.getUnverifiedReasons());
+            for (Record<? extends Data> record : dnskeyRes.answerSection) {
+                Record<DNSKEY> dnsKeyRecord = record.ifPossibleAs(DNSKEY.class);
+                if (dnsKeyRecord == null) continue;
+
+                if (dnsKeyRecord.payloadData.getKeyTag() == rrsig.keyTag) {
+                    dnskey = dnsKeyRecord.payloadData;
+                }
+            }
+        }
+        if (dnskey == null) {
+            throw new DNSSECValidationFailedException(q, records.size() + " " + rrsig.typeCovered + " record(s) are signed using an unknown key.");
+        }
+        UnverifiedReason unverifiedReason = verifier.verify(records, rrsig, dnskey);
+        if (unverifiedReason != null) {
+            result.add(unverifiedReason);
+        }
+        return result;
+    }
+
+    private Set<UnverifiedReason> verifySecureEntryPoint(Question q, final Record<DNSKEY> sepRecord) throws IOException {
+        final DNSKEY dnskey = sepRecord.payloadData;
+
+        Set<UnverifiedReason> unverifiedReasons = new HashSet<>();
+        Set<UnverifiedReason> activeReasons = new HashSet<>();
+        if (knownSeps.containsKey(sepRecord.name)) {
+            if (dnskey.keyEquals(knownSeps.get(sepRecord.name))) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **385 and 386**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The removes that can remove the key are at lines: **500** and **510**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).